src/java/org/apache/cassandra/utils/logging/AbstractVirtualTableAppender.java

```diff
@@ -0,0 +1,135 @@
+                                                      .filter(vt -> vt.name().equals(tableName))
+                                                      .findFirst();
+
+        if (virtualTable.isEmpty())
+            return null;
+
+        VirtualTable vt = virtualTable.get();
+
+        if (!vt.getClass().equals(vtableClass))
+            throw new IllegalStateException(String.format("Virtual table %s.%s is not backed by an instance of %s but by %s",
+                                                          VIRTUAL_VIEWS,
+                                                          tableName,
+                                                          vtableClass.getName(),
+                                                          vt.getClass().getName()));
+
+        return (T) vt;
+    }
+
+    /**
+     * This method adds an event to virtual table, when present.
+     * When vtable is null, we will attempt to find it among registered ones. Then not found, we add it to internal
+     * buffer for later processing. This might happen e.g. for logging tables when log events
+     * were appended via logging framework sooner than registration of virtual tables was done so after they are registered,
+     * they would miss logging events happened before being so.
+     *
+     * @param vtable    vtable to append to
+     * @param event     event to append to
+     * @param tableName table name of virtual table to append to
+     * @return vtable or when null, found vtable
+     */
+    protected AbstractLoggerVirtualTable<?> appendToVirtualTable(AbstractLoggerVirtualTable<?> vtable, LoggingEvent event, String tableName)
+    {
+        AbstractLoggerVirtualTable<?> foundVtable;
+        if (vtable == null)
+        {
+            foundVtable = getVirtualTable(SlowQueriesTable.class, tableName);
+            if (foundVtable == null)
+                addToBuffer(event);
+            else
+                foundVtable.add(event);
+        }
+        else
+        {
+            foundVtable = vtable;
+            vtable.add(event);
+        }
+
+        return foundVtable;
+    }
+
+    @Override
+    public void stop()
+    {
+        messageBuffer.clear();
+        super.stop();
+    }
+
+    /**
+     * Flushes all log entries which were appended before virtual table was registered.
+     *
+     * @see org.apache.cassandra.service.CassandraDaemon#setupVirtualKeyspaces
+     */
+    public void flushBuffer(Class<? extends AbstractLoggerVirtualTable<?>> vtableClass, String tableName)
+    {
+        Optional.ofNullable(getVirtualTable(vtableClass, tableName)).ifPresent(vtable -> {
+            messageBuffer.forEach(vtable::add);
+            messageBuffer.clear();
+        });
+    }
+
+    protected void addToBuffer(LoggingEvent eventObject)
+    {
+        // we restrict how many logging events we can put into buffer,
+        // so we are not growing without any bound when things go south
+        if (messageBuffer.size() < defaultRows)
```

Comment:
Should we add a logger debug if message is not added to the buffer due to buffer size?

Reply:
That would most probably caused a flood of debug messages every time we are "over the limit" unless the logger is no-spamming one. I would keep it like it is.